parsers/julia.c

```diff
@@ -1360,5 +1383,6 @@ extern parserDefinition* JuliaParser (void)
     def->parser     = findJuliaTags;
     def->keywordTable = JuliaKeywordTable;
     def->keywordCount = ARRAY_SIZE (JuliaKeywordTable);
+    def->useCork    = CORK_QUEUE;
```

Comment:
I don't think I understand cork correctly. From the doc and the clojure parser, it seems the way to use cork is make a tag for the scope/namespace first, then set a tag's `scopeIndex` to it. But we don't want to make a tag for `Base` in a block like

``` julia
function Base.func
    something...
end
```

The Ruby parser uses cork, but doesn't do this, and doesn't use `scopeIndex`.

Besides, setting `requestAutomaticFQTag` doesn't give me full qualified tags for Julia, so I must be using cork in a wrong way.

Reply:
Julia parser, including your changes, manages scope information (both names and kinds manually) manually.
WIth the cork_queue, your parser becomes free from tracking them.
With the cork_queue, your parser can track only an integer (cork index) returned from makeTagEntry.
The cork index is a candidate for the scope of another tag.

As usual, I will show the way how to use the cork queue.
What you have to do first of all is:
```diff
diff --git a/parsers/julia.c b/parsers/julia.c
index ac7df5e5..3b17ecb7 100644
--- a/parsers/julia.c
+++ b/parsers/julia.c
@@ -823,11 +823,11 @@ static void debugLexer (lexerState *lexer)
 }
 #endif
 
-static void addTag (vString* ident, const char* type, const char* arg_list, int kind, unsigned long line, MIOPos pos, vString *scope, int parent_kind)
+static int addTag (vString* ident, const char* type, const char* arg_list, int kind, unsigned long line, MIOPos pos, vString *scope, int parent_kind)
 {
     if (kind == K_NONE)
     {
-        return;
+        return CORK_NIL;
     }
     tagEntryInfo tag;
     initTagEntry(&tag, vStringValue(ident), kind);
@@ -843,7 +843,7 @@ static void addTag (vString* ident, const char* type, const char* arg_list, int
         tag.extensionFields.scopeKindIndex = parent_kind;
         tag.extensionFields.scopeName = vStringValue(scope);
     }
-    makeTagEntry(&tag);
+    return makeTagEntry(&tag);
 }
 
 /* Skip tokens until one of the goal tokens is hit. Escapes when level = 0 if there are no goal tokens.
@@ -1360,5 +1360,6 @@ extern parserDefinition* JuliaParser (void)
     def->parser     = findJuliaTags;
     def->keywordTable = JuliaKeywordTable;
     def->keywordCount = ARRAY_SIZE (JuliaKeywordTable);
+	def->useCork = CORK_QUEUE;
     return def;
 }
```

You can use the above change with code like `int parent_index = addTag ("PARENT"...);`.
The next step is for providing the way to fill a scope field for a new tag:
```diff
diff --git a/parsers/julia.c b/parsers/julia.c
index 3b17ecb7..048f6980 100644
--- a/parsers/julia.c
+++ b/parsers/julia.c
@@ -823,7 +823,8 @@ static void debugLexer (lexerState *lexer)
 }
 #endif
 
-static int addTag (vString* ident, const char* type, const char* arg_list, int kind, unsigned long line, MIOPos pos, vString *scope, int parent_kind)
+static int addTag (vString* ident, const char* type, const char* arg_list, int kind, unsigned long line, MIOPos pos, vString *scope, int parent_kind,
+	int parent_represented_in_corkIndex)
 {
     if (kind == K_NONE)
     {
@@ -838,7 +839,9 @@ static int addTag (vString* ident, const char* type, const char* arg_list, int k
 
     tag.extensionFields.signature = arg_list;
     /* tag.extensionFields.varType = type; */  /* Needs a workaround */
-    if (parent_kind != K_NONE)
+	if (parent_represented_in_corkIndex != CORK_NIL)
+		tag.extensionFields.scopeIndex = parent_represented_in_corkIndex;
+	else if (parent_kind != K_NONE)
     {
         tag.extensionFields.scopeKindIndex = parent_kind;
         tag.extensionFields.scopeName = vStringValue(scope);
```

When you want to make a tag for "CHILD" having "PARENT" as the scope, call addTag like `addTag("CHILD", ..., parent_index, ...);`.  If you don't want to fill scope the original way, pass `CORK_NIL` instead of `parent_index`.

Reply:
In some cases, you may want to make X as a scope without tagging X. Anonymous tag feature and reference tags may help you in that case.

Reply:
Thanks!

> In some cases, you may want to make X as a scope without tagging X.

Yes, that's the main consideration. I do saw something like `extras:anonymous` in some tags files. I think I've got the idea.

I'll come back in 1 or 2 days.

Reply:
If there is no name for the scope, the anonymous tag is for you.

input.c:
```c
struct {
  int x, y;
} origin;
```
u-ctags -o - /tmp/input.c
```
__anon6dc190750108	/tmp/input.c	/^struct {$/;"	s	language:C	file:
origin	/tmp/input.c	/^} origin;$/;"	v	language:C	typeref:struct:__anon6dc190750108
x	/tmp/input.c	/^  int x, y;$/;"	m	language:C	struct:__anon6dc190750108	typeref:typename:int	file:
y	/tmp/input.c	/^  int x, y;$/;"	m	language:C	struct:__anon6dc190750108	typeref:typename:int	file:
```

For making anonymous tags, you can use anon* functions declared in main/parse.h.
Don't forget to call `markTagExtraBit (&e, XTAG_ANONYMOUS);` after making an anonymous tag.

If the scope has a name but it is not a definition, the reference tag is for you.
input.py:
```python
from x import y
```
u-ctags --fields=+E --extras=+r -o - /tmp/input.py 
```
x	/tmp/input.py	/^from x import y$/;"	i	language:Python	extras:reference
y	/tmp/input.py	/^from x import y$/;"	x	language:Python	module:x	extras:reference
```

Reply:
I gave up the idea of using cork, mainly because the julia parser handles scope information in its own way. The code becomes awkward if we mix them up, and it's beyond my ability to fully replace it using cork.

But thanks for your help anyway. Maybe I could use them to improve the documentation of cork.